### PR TITLE
T31627 - create lava job for chrome os

### PR DESCRIPTION
**Pull request by alexandrasp**
Add a new lava job template to be used by chrome OS at this first step to run Tast tests.

### CODE REVIEW

File: config/lava/boot/generic-emmc-tftp-chromebook-boot-template.jinja2 [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{% extends 'base/kernel-ci-base-tftp-deploy.jinja2' %}
```
Comment (gctucker) [inline config/lava/boot/generic-emmc-tftp-chromebook-boot-template.jinja2:1]: That won't work because of the modules

File: config/core/test-configs.yaml [hunk 2]
```diff
@@ -357,6 +368,12 @@ test_plans:
   smc:
     rootfs: buildroot_baseline_ramdisk
 
+  tast:
+    rootfs: chrome_version
+    pattern: 'chromeos/new-chromeos-template.jinja2'
+    filters:
+      - passlist: {defconfig: ['x86_64_defconfig+x86-chromebook']}
```
Comment (gctucker) [inline config/core/test-configs.yaml:375]: This is specific to devices, not to the test plan

File: config/core/test-configs.yaml [hunk 3]
```diff
@@ -42,6 +47,12 @@ file_systems:
     nfs: 'qemu-{arch}/cip-core-image-kernelci-cip-core-buster-qemu-{arch}.tar.gz'
     root_type: nfs
 
+  chrome_version:
+    type: chrome-os
+    ramdisk: 'to be defined'
+    emmc: 'to be defined'
+    root_type: emmc
```
Comment (gctucker) [inline config/core/test-configs.yaml:54]: We can't add the rootfs image for now, that will be done by the part of the job that does the flashing

File: config/core/test-configs.yaml [hunk 4]
```diff
@@ -357,6 +368,12 @@ test_plans:
   smc:
     rootfs: buildroot_baseline_ramdisk
 
+  tast:
+    rootfs: chrome_version
```
Comment (gctucker) [inline config/core/test-configs.yaml:372]: This can be dropped

File: config/core/test-configs.yaml [hunk 5]
```diff
@@ -357,6 +368,12 @@ test_plans:
...
+    pattern: 'chromeos/new-chromeos-template.jinja2'
```
Comment (gctucker) [inline config/core/test-configs.yaml:373]: This name doesn't seem better than `chromeos/tast.jinja2`

File: config/lava/chromeos/chromeos.jinja2 [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            - ls -l /home/cros-tast
+#             - 'cd /home/cros-tast; curl -s http://172.17.0.1/chromeos/autotest_server_package_octopus_R92_13981.0.0.tar.bz2 | tar xjf -'
+#             - 'chown -R cros-tast: /home/cros-tast'
+            - ls -l /home/cros-tast
+            - ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i /home/cros-tast/.ssh/id_rsa root@$(lava-target-ip) whoami
+#             - 'cd /home/cros-tast/tast; ./tast list -build=false root@$(lava-target-ip)'
+#             - while true; do date; sleep 15; done
+      from: inline
+      name: chromeos-tast-inline
+      path: inline/chromeos-tast.yaml
```
Comment (gctucker) [inline config/lava/chromeos/chromeos.jinja2:32]: Please drop the debug stuff

File: config/lava/chromeos/chromeos.jinja2 [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    - repository: tast tarball?
+      name: {{ specific_tast_test }}
```
Comment (gctucker) [inline config/lava/chromeos/chromeos.jinja2:35]: The tarball is not with test definitions but with the Tast framework binaries, so this should be dropped

**Review by gctucker — CHANGES_REQUESTED** [review]
The PR should be for the `chromeos` branch rather than `main`, and there are several new issues compared to the draft I made yesterday.

File: config/lava/boot/generic-emmc-tftp-chromebook-boot-template.jinja2 (hunk 1 above)
Comment (alexandrasp) [reply config/lava/boot/generic-emmc-tftp-chromebook-boot-template.jinja2]: updated

File: config/core/test-configs.yaml (hunk 2 above)
Comment (alexandrasp) [reply config/core/test-configs.yaml]: updated

File: config/core/test-configs.yaml (hunk 3 above)
Comment (alexandrasp) [reply config/core/test-configs.yaml]: updated

File: config/core/test-configs.yaml (hunk 5 above)
Comment (alexandrasp) [reply config/core/test-configs.yaml]: updated

File: config/lava/chromeos/chromeos.jinja2 (hunk 6 above)
Comment (alexandrasp) [reply config/lava/chromeos/chromeos.jinja2]: updated

File: config/lava/chromeos/chromeos.jinja2 (hunk 7 above)
Comment (alexandrasp) [reply config/lava/chromeos/chromeos.jinja2]: updated

File: config/core/test-configs.yaml [hunk 8]
```diff
@@ -691,12 +694,14 @@ device_types:
     filters:
       - passlist: {defconfig: ['x86-chromebook']}
 
-  hp-x360-12b-n4000-octopus:
+  hp-x360-12b-n4000-octopus: &n4000-octopus
```
Comment (gctucker) [inline config/core/test-configs.yaml:697]: Please move this to a different PR.  Also I think we need to review the octopus device type names as we may not need to keep the 2 variants.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      prompts:
+        - 'sof-audio-pci'  # ToDo: make a parameter
+
+  - test:
+      timeout:
+        minutes: 45
+      docker:
+        image: kernelci/chromeos-tast  # ToDo: make a parameter
+        volumes:
+          - /tmp/octopus:/home/cros-tast/octopus  # ToDo: make a parameter
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:32]: I don't think this is actually used, as the tarball is downloaded over HTTP from the host.

**Review by gctucker — CHANGES_REQUESTED** [review]
Could you please add links to LAVA jobs generated and submitted from this template when you have some?

**Comment by alexandrasp** [comment]
> Could you please add links to LAVA jobs generated and submitted from this template when you have some?

yes! Yesterday I tried to submit some jobs, but I am not sure about what is going on, it seems that I don't have permission..

https://lava.collabora.co.uk/scheduler/job/5533790
https://staging.lava.collabora.dev/scheduler/job/100670

I tried to run a random job submitted by you just to check and nothing happens again
https://lava.collabora.co.uk/scheduler/job/5537397

**Comment by gctucker** [comment]
The permission issue with the main lab is being investigated.  On staging, that's because the device dictionary was reverted to the standard one which doesn't have the `emmc` boot commands.

@mgalka Could you please put back the custom device dictionary on the staging octopus?

File: config/core/test-configs.yaml [hunk 10]
```diff
@@ -137,6 +137,16 @@ test_plans:
     filters:
       - blocklist: *kselftest_defconfig_filter
 
+  chromeos-tast:
+    pattern: 'chromeos/chromeos-tast-template.jinja2'
+    params:
+      job_timeout: '60'
+      prompt_command: 'sof-audio-pci'
+      docker_image: 'kernelci/chromeos-tast'
+      chromeos_volume: '/tmp/octopus:/home/cros-tast/octopus'
```
Comment (gctucker) [inline config/core/test-configs.yaml:146]: I don't think we need this

File: config/core/test-configs.yaml [hunk 11]
```diff
@@ -137,6 +137,16 @@ test_plans:
...
+      cros_tast_tarball: 'http://172.17.0.1/chromeos/autotest_server_package_octopus_R92_13981.0.0.tar.bz2'
+      kernel_url: 'https://images.collabora.com/lava/chromeos/vmlinuz-4.14.243'
```
Comment (gctucker) [inline config/core/test-configs.yaml:148]: The kernel should not be defined here, it's the moving part in each test.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      prompts:
+        - {{ prompt_command }}
+
+  - test:
+      timeout:
+        minutes: 45
+      docker:
+        image: {{ docker_image }}
+        volumes:
+          - {{ chromeos_volume }}
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:32]: This can be dropped

**Review by gctucker — COMMENTED** [review]
Please upload a kernel image to people.collabora.com or storage.staging.kernelci.org with the JSON meta-data adjusted so that kci_test can use it to generate and submit a LAVA job using this template.

**Comment by gctucker** [comment]
> Please upload a kernel image to people.collabora.com or storage.staging.kernelci.org with the JSON meta-data adjusted so that kci_test can use it to generate and submit a LAVA job using this template.

@alexandrasp Any luck with this?

**Comment by alexandrasp** [comment]
> > Please upload a kernel image to people.collabora.com or storage.staging.kernelci.org with the JSON meta-data adjusted so that kci_test can use it to generate and submit a LAVA job using this template.
> 
> @alexandrasp Any luck with this?

yes, I had a problem to access people, sysadmins solved today! 
Link to a job using kernel image from my people page : https://staging.lava.collabora.dev/scheduler/job/104411

**Comment by gctucker** [comment]
Great. Also I see listing is not enabled in this directory:
https://people.collabora.com/~alex.pereira/chromeos-images/

Which is the default setting, but if you want you can add a file like this there:
```
$ cat .htaccess 
Options +Indexes
```
Then the list of files (index) will be shown when visiting the directory.

**Comment by gctucker** [comment]
The job definition is missing the extra kernel arguments with the root partition and the cros_* flags.  See this sample job:
https://staging.lava.collabora.dev/scheduler/job/101368/definition

**Comment by gctucker** [comment]
Also please rebase your branch, I rebased `chromeos` yesterday on top of `main`.

**Comment by alexandrasp** [comment]
> @alexandrasp Please rebase your branch so only your new commits appear in this PR.

done, I got lost between chrome and main branches.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 13]
```diff
@@ -0,0 +1,45 @@
+{% extends 'base/kernel-ci-base.jinja2' %}
+{% include "/chromeos/chromeos-extra-args.jinja2" %}
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:2]: Does this need to be in a separate file? Can it just be defined in this file directly?

File: config/core/test-configs.yaml [hunk 14]
```diff
@@ -756,6 +764,7 @@ device_types:
     mach: x86
     arch: x86_64
     boot_method: depthcharge
+    context: {extra_kernel_args: cros_debug cros_secure root=/dev/mmcblk0p3}
```
Comment (alexandrasp) [inline config/core/test-configs.yaml:767]: @gctucker I am not sure if we can keep this context here..

Using [artifact and bmeta json files](https://people.collabora.com/~alex.pereira/chromeos/chromeos-images/) availables on my page it is possible to build LAVA job template using `kci_test` and run it in a lab.

Last test results:
https://lava.collabora.co.uk/scheduler/job/5763022

Comment (gctucker) [reply config/core/test-configs.yaml:767]: Yes, these parameters should be moved to the chromeos-tast test plan definition. Maybe it could be added to the `context` Jinja2 dictionary variable in `chromeos-tast-template.jinja2`? I think it's fine to hard-code it in this template since it's going to be the same for all Chrome OS tests.

Comment (gctucker) [reply config/core/test-configs.yaml:767]: To check your changes, you could also generate and submit a baseline test for the same device and same kernel meta-data etc.

Comment (alexandrasp) [reply config/core/test-configs.yaml:767]: > Yes, these parameters should be moved to the chromeos-tast test plan definition. Maybe it could be added to the `context` Jinja2 dictionary variable in `chromeos-tast-template.jinja2`? I think it's fine to hard-code it in this template since it's going to be the same for all Chrome OS tests.

The first strategy was adding context in template hard-coded, but..it will be a break in "kernel CI design" usually we define variables in test-configs. :)

**Comment by alexandrasp** [comment]
Re-tested after move `context` to test plan.

Baseline job: https://lava.collabora.co.uk/scheduler/job/5766468
ChromeOS job: https://lava.collabora.co.uk/scheduler/job/5766469

File: config/core/test-configs.yaml (hunk 14 above)
Comment (gctucker) [reply config/core/test-configs.yaml]: How about devices that also define a `context` template parameter though? I think they're going to collide and only one of them will actually be in the template.

For example:
```yaml
  bcm2837-rpi-3-b:
    mach: broadcom
    class: arm64-dtb
    boot_method: uboot
    context: {console_device: 'ttyS1', test_character_delay: 10}
    filters:
      - blocklist:
          defconfig: *device_config_filter
          kernel: ['v3.', '4.4-', '4.9-', 'v4.13']
```
That's why I thought, if the Chrome OS template updates the `context` dictionary then things will get merged together.

See the code in `kernelci.test`:
```python
def get_params(meta, target, plan_config, storage):
    # ...
    if rootfs:
        # ...
        params.update(rootfs.params)

    params.update(plan_config.params)
    params.update(target.params)


    return params
```
So we would need another PR to replace the plain `.update()` calls with a function to recursively update dictionaries inside each config params.

I think that can be done and it would seem like a good improvement, but maybe we can get this PR merged by hard-coding the parameters in the template and then improve it as a follow-up.

**Review by gctucker — APPROVED** [review]

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 15]
```diff
@@ -0,0 +1,46 @@
+{% extends 'base/kernel-ci-base.jinja2' %}
+{% block actions %}
+context: 
+  extra_kernel_args: cros_debug cros_secure root=/dev/mmcblk0p3
```
Comment (alexandrasp) [inline config/lava/chromeos/chromeos-tast-template.jinja2:4]: hard-code here, I will work as you suggest in a follow up improvement.

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2:4]: I meant, using the `context.update()` method on the template variable so that it gets added to any other context.

See this for example:
```
config/lava/boot/generic-depthcharge-tftp-ramdisk-boot-template.jinja2:{% do context.update({"extra_kernel_args": extra_kernel_args}) %}
```

Which is by default set here:
```
config/lava/base/kernel-ci-base.jinja2:{% set extra_kernel_args = "console_msg_format=syslog earlycon" %}
```

So in this template, you could define this variable:
```
{% set extra_kernel_args = "console_msg_format=syslog earlycon" cros_debug cros_secure root=/dev/mmcblk0p3 %}
```

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 16]
```diff
@@ -0,0 +1,45 @@
+{% extends 'base/kernel-ci-base.jinja2' %}
+{% block actions %}
+{% set extra_kernel_args = "console_msg_format=syslog earlycon" cros_debug cros_secure root=/dev/mmcblk0p3 %}
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:3]: There's a problem with the double-quotes there.

Comment (alexandrasp) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: finally it works again..
New lava job to test: https://lava.collabora.co.uk/scheduler/job/5794516

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: Well, the job above landed on octopus cbg-2 which doesn't have a Chrome OS image install. So that's why it failed. If you resubmit it with a tag set to cbg-0 or cbg-1 it should work.

Also someone should flash all the octopus devices with a Chrome OS image to avoid this problem, ideally with a job kept under source control. @mgalka Any news on this front?

Comment (alexandrasp) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: cbg-0 : https://lava.collabora.co.uk/scheduler/job/5801891
cbg-1: https://lava.collabora.co.uk/scheduler/job/5801894

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: Great. There's a problem with the Chrome OS setup on these devices, but that's a known issue:
```
Process exited with status 127 (output: \"bash: line 0: exec: file: not found\n\")
```
That's because /usr/local is not mounted. It's not an issue with this PR, so I think it's working as expected.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 17]
```diff
@@ -0,0 +1,51 @@
+{% extends 'base/kernel-ci-base.jinja2' %}
+{% block main %}
+{%- if extra_kernel_args %}
```
[the remark this answers is not in the document]
Comment (alexandrasp) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: not reason to use this `if`, removed.

File: config/lava/chromeos/chromeos-tast-template.jinja2 [hunk 18]
```diff
@@ -0,0 +1,49 @@
+{% extends 'base/kernel-ci-base.jinja2' %}
+{% block main %}
+{% set extra_kernel_args = "console_msg_format=syslog earlycon cros_debug cros_secure root=/dev/mmcblk0p3" %}
+{% do context.update({"extra_kernel_args": extra_kernel_args}) %}
```
Comment (gctucker) [inline config/lava/chromeos/chromeos-tast-template.jinja2:4]: Oh I see, the `kernel-ci-base.jinja2` template doesn't do `context.update()`. So we could just have this here:
```suggestion
{% do context.update({"extra_kernel_args": "console_msg_format=syslog earlycon cros_debug cros_secure root=/dev/mmcblk0p3" }) %}
```
But that's fine, as eventually we should be using a `extra_kernel_args` variable. So we can merge it as-is I think.

Comment (alexandrasp) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: yes, the `kernel-ci-base.jinja2` template doesn't have context update :/ but I've updated considering your suggestion. I will open a PR to do parameters a bit more flexible as you suggested above. :)

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: Have you submitted a job with the latest version of this PR? If so, please provide a LAVA job link to keep a trace.

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: And it failed, but where is this kernel image coming from?
```yaml
      kernel:
        url: https://people.collabora.com/~alex.pereira/chromeos/chromeos-images/chromeos-20220209-4.14.243
```
There's a good known 4.14 Chrome OS kernel for octopus so we should be using that while setting things up in LAVA.

Comment (gctucker) [reply config/lava/chromeos/chromeos-tast-template.jinja2]: Resubmitted it with the working kernel URL and that worked:
https://lava.collabora.co.uk/scheduler/job/5817798
(except the issue of /usr/local not being mounted but that's not an issue with this PR).

So I think this is ready to be merged.  I'd just like to get to the bottom of the kernel being used in the first job first.
